pkg/image/registry/imagestreamtag/clone.go

```diff
@@ -0,0 +1,793 @@
+	imageClone, ok := obj.(*imageapi.ImageStreamTagClone)
+	if !ok {
+		return nil, errors.NewBadRequest(fmt.Sprintf("obj is not an ImageStreamTagClone: %#v", obj))
+	}
+	uid := imageClone.UID
+	imageClone.UID = ""
+	if err := rest.BeforeCreate(CloneStrategy, ctx, obj); err != nil {
+		return nil, err
+	}
+
+	// TODO: decide whether specifying UID precondition like this is acceptable
+	var preconditionUID *types.UID
+	if len(uid) > 0 {
+		preconditionUID = &imageClone.UID
+	}
+
+	id := imageClone.Name
+	imageStreamName, _, ok := imageapi.SplitImageStreamTag(id)
+	if !ok {
+		return nil, fmt.Errorf("%q must be of the form <stream_name>:<tag>", id)
+	}
+	namespace, ok := apirequest.NamespaceFrom(ctx)
+	if !ok {
+		return nil, errors.NewBadRequest("a namespace must be specified to clone images")
+	}
+
+	target, tag, err := imageStreamForClone(r.imageStreamRegistry, ctx, id, r.gr, preconditionUID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+
+		// try to create the target if it doesn't exist
+		target = &imageapi.ImageStream{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      imageStreamName,
+				Namespace: namespace,
+			},
+		}
+	}
+
+	// if target.Spec.Tags == nil {
+	// 	target.Spec.Tags = make(map[string]imageapi.TagReference)
+	// }
+
+	return r.completeClone(ctx, tag, target, imageClone, nil, "")
+}
+
+func (r *CloneREST) completeClone(ctx apirequest.Context, tag string, target *imageapi.ImageStream, imageClone *imageapi.ImageStreamTagClone, layerBody io.Reader, mediaType string) (runtime.Object, error) {
+	// TODO: load this from the default registry function
+	insecure := true
+
+	ref, u, err := registryTarget(target, r.defaultRegistry)
+	if err != nil {
+		return nil, err
+	}
+
+	// verify the user has access to the From image, if any is specified
+	baseImageName, baseImageRepository, err := r.resolveTagCloneToImage(ctx, target, imageClone)
+	if err != nil {
+		return nil, err
+	}
+
+	// no layer, so we load our base image (if necessary)
+	var baseImage *imageapi.Image
+	var sourceRepo distribution.Repository
+	if len(baseImageName) > 0 {
+		image, err := r.imageRegistry.GetImage(ctx, baseImageName, &metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		baseImage = image
+		sourceRepo, err = r.repository.Repository(ctx, u, baseImageRepository, insecure)
+		if err != nil {
+			return nil, errors.NewInternalError(fmt.Errorf("could not contact integrated registry: %v", err))
+		}
+		glog.V(4).Infof("Using base image for clone of tag %s: %s from %s", imageClone.Name, baseImageName, baseImageRepository)
+	}
+
+	imageRepository := imageapi.DockerImageReference{Namespace: ref.Namespace, Name: ref.Name}.Exact()
+	repo, err := r.repository.Repository(ctx, u, imageRepository, insecure)
+	if err != nil {
+		return nil, errors.NewInternalError(fmt.Errorf("could not contact integrated registry: %v", err))
+	}
+
+	var imageLayer *imageapi.ImageLayer
+	var imageLayerDiffID digest.Digest
+	if r != nil {
```

Comment:
Yeah, should probably be renamed when we're in a context where we are using it as a specific type.